### PR TITLE
Feature/6.dev/migrations

### PR DESCRIPTION
**Pull request by matthewjohns0n**
## Overview

Adds full featured CLI with migrations to the core.

## Nature of This Change

- [ ] 🐛 Fixes a bug
- [x] 🚀 Implements a new feature
- [ ] 🛁 Refactors existing code
- [ ] 💅 Fixes coding style
- [ ] ✅ Adds tests
- [ ] 👽 Adds new dependency
- [ ] 🔥 Removes unused files / code
- [ ] 🔒 Improves security 

## Is this backwards compatible?

- [x] Yes
- [ ] No

## Documentation

User Guide Pull Request: https://github.com/ExpressionEngine/ExpressionEngine-User-Guide/pulls/NNN

### CODE REVIEW

File: system/ee/ExpressionEngine/Cli/Commands/CommandMakeAddon.php [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+        // If they passed the settings flag, always take that
+        if ($this->option('--has-settings')) {
+            $this->data['has_settings'] = $this->option('--has-settings');
+        } elseif ($this->data['type'] == 'plugin') {
+            // Default to no if it's a plugin
+            $this->data['has_settings'] = 'no';
+        } else {
+            // Ask if not passed and not a plugin
+            $this->data['has_settings']  = $this->confirm("Does your {$this->type['slug']} " . lang('command_make_addon_have_settings_question'));
```
Comment (obfuscode) [inline system/ee/ExpressionEngine/Cli/Commands/CommandMakeAddon.php]: Is the mix of this English and lang file text an issue?

File: system/ee/ExpressionEngine/Cli/Commands/CommandMakeAddon.php [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+        return $type;
+    }
+
+    private function getName()
+    {
+        $name = $this->getFirstUnnamedArgument();
+
+        if (is_null($name)) {
+            $name = $this->ask("What is the name of your add-on?");
```
Comment (obfuscode) [inline system/ee/ExpressionEngine/Cli/Commands/CommandMakeAddon.php]: Does this need to be extracted to the lang file?

File: system/ee/ExpressionEngine/Cli/Commands/Migration/Templates/AbstractTemplate.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        $missing = array();
+        foreach ($this->requiredVars() as $requiredVar) {
+            if (! array_key_exists($requiredVar, $this->vars)) {
+                $missing[] = $requiredVar;
+            }
+        }
+
+        // If we found a required variable that is missing, let the user know
+        if (!empty($missing)) {
+            echo "Missing required variable for parsing migration template:\n";
```
Comment (obfuscode) [inline system/ee/ExpressionEngine/Cli/Commands/Migration/Templates/AbstractTemplate.php]: Needs language translation

**Review by obfuscode — CHANGES_REQUESTED** [review]
Needs to extract some language strings.

File: system/ee/ExpressionEngine/Cli/Commands/Migration/Templates/AbstractTemplate.php (hunk 3 above)
Comment (matthewjohns0n) [reply system/ee/ExpressionEngine/Cli/Commands/Migration/Templates/AbstractTemplate.php]: This has been updated in 76c29552d09d443e697d78e03414032dfa324eb3.

**Review by obfuscode — APPROVED** [review]

**Pull request merged**